### PR TITLE
[move-lang] Specification syntax changes

### PR DESCRIPTION
**Pull request by wrwg**
[Note: because the larger number of changes in .move sources, we should try to land this fast and iterate.]

This PR implements a number of changes to the specification syntax:

- We not longer write `spec fun name` or `spec struct name`, but just `spec name`. What it is will be determined by name resolution context.
- As `spec fun` is now free, we can use it for specifications functions. So instead of `spec define ..` we write `spec fun ..` (respectively, instead of `spec module { define ..` write `spec module { fun ..`),
- We can now put specifications into separate "specification modules". These have the form `spec 0x1::Name` and appear on the same level as modules. They can also be put into separate files if desired. An example is [here](https://github.com/wrwg/diem/blob/spec-syn/language/move-prover/tests/sources/functional/module_level_spec.move).
- We can now directly write `invariant` on module level for a global invariant. 
- We can now provide a signature for a member specification. For example, we can say `spec foo_fun(x: u64): u64 { <specs> }` to give context to a function spec, specifically if it lives further away from the definition. This feature is optional.

The PR also removes some older not longer supported specification language features, namely `invariant module`, `invariant pack`, etc. The property `[global]` is not longer required for invariants on module level.

Sources have been upgraded based on a new script in `move-prover/tools/migrate_soec_fun_syntax.sh`.

## Motivation

Nicer syntax.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Existing tests + new tests

## Related PRs

NA

### CODE REVIEW

File: language/move-lang/src/parser/merge_spec_modules.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        .filter_map(|def| match def {
+            Module(m) => extract_spec_module(&None, m, spec_modules).map(Module),
+            Address(attrs, loc, addr, module_defs) => {
+                let addr_opt = Some(addr);
+                let module_defs = module_defs
+                    .into_iter()
+                    .filter_map(|m| extract_spec_module(&addr_opt, m, spec_modules))
+                    .collect::<Vec<_>>();
+                if module_defs.is_empty() {
+                    None
```
Comment (tnowacki) [inline language/move-lang/src/parser/merge_spec_modules.rs:65]: It is fine to have an empty address decl, especially once named address syntax is in, so the address should also be kept

**Review by tzakian — COMMENTED** [review]
Had a look at the Move changes and they look good to me.

Liking the new syntax!

File: language/move-lang/src/parser/syntax.rs [hunk 2]
```diff
@@ -1930,13 +1962,39 @@ fn parse_module(
...
+                    "constants not allowed in specification module".to_owned(),
+                )),
+                ModuleMember::Use(_) | ModuleMember::Friend(_) | ModuleMember::Spec(_) => None,
+            })
+            .collect();
+        if !errors.is_empty() {
+            return Err(errors);
+        }
+        // (b) add implicit uses
+    }
```
[the remark this answers is not in the document]
Comment (wrwg) [reply language/move-lang/src/parser/syntax.rs]: Yes. Done.

File: language/move-lang/src/parser/syntax.rs [hunk 3]
```diff
@@ -1861,7 +1861,7 @@ fn parse_use_alias(tokens: &mut Lexer) -> Result<Option<Name>, Error> {
 
 // Parse a module:
 //      Module =
-//          <DocComments> "module" (<Address>::)?<ModuleName> "{"
+//          <DocComments> ["spec"] "module" (<Address>::)?<ModuleName> "{"
```
[the remark this answers is not in the document]
Comment (wrwg) [reply language/move-lang/src/parser/syntax.rs]: Left over comment, fixed.

File: language/move-lang/src/parser/syntax.rs [hunk 4]
```diff
@@ -1893,7 +1899,33 @@ fn parse_module(
...
+                    tokens.match_doc_comments();
+                    ModuleMember::Spec(singleton_module_spec_block(
+                        tokens,
+                        tokens.start_loc(),
+                        attributes,
+                        parse_invariant,
+                    )?)
+                }
+                Tok::Spec => {
+                    if matches!(tokens.lookahead()?, Tok::Fun | Tok::Native) {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply language/move-lang/src/parser/syntax.rs]: Done

File: language/move-lang/src/parser/merge_spec_modules.rs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+fn merge_spec_module(
+    address_opt: &Option<Address>,
+    m: &mut ModuleDefinition,
+    spec_modules: &mut BTreeMap<(Address, String), ModuleDefinition>,
+) {
+    if let Some(spec_module) = spec_modules.remove(&module_key(address_opt, m)) {
+        let ModuleDefinition {
+            attributes,
+            members,
+            ..
```
[the remark this answers is not in the document]
Comment (wrwg) [reply language/move-lang/src/parser/merge_spec_modules.rs:139]: Other fields or name, loc, etc. and not empty.

File: language/move-lang/src/parser/merge_spec_modules.rs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        } = spec_module;
+        m.attributes.extend(attributes.into_iter());
+        m.members.extend(members.into_iter());
+    }
+}
+
+fn module_key(address_opt: &Option<Address>, def: &ModuleDefinition) -> (Address, String) {
+    let name = def.name.0.value.clone();
+    if let Some(a) = def.address {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply language/move-lang/src/parser/merge_spec_modules.rs]: Done

File: language/move-lang/src/parser/merge_spec_modules.rs (hunk 1 above)
Comment (wrwg) [reply language/move-lang/src/parser/merge_spec_modules.rs]: Done

**Review by tnowacki — APPROVED** [review]
Thanks for making these syntax changes! Have had them on my mind for a long time

It's a huge PR so let's get it in, other things can be cleaned up if needed later

**Pull request merged**